Fixes #37293 - Add customizable columns to new Hosts index page

jeremylenz opened the pull request:
This PR adds customizable columns to the React (experimental) HostsIndex page.

1.  Respects the columns already saved in the current user's `TablePreference` for the `'hosts'` table, and uses that to decide which columns to display.
2. Columns that can be displayed are defined via a registry, so that plugins can also contribute columns. See the developer documentation for more details.
3. Existing data for most columns was copied over from the Ruby pagelets to React/JSX. Note that there were a few columns such as Power status and Virtual that are not currently part of the API response, so I'm still thinking about how to handle those. I may leave those for a follow-up.

Note that this pattern can now be used with _any_ React table in Foreman that uses TableIndex, TablePreferences, and the ColumnSelector component. In this way we can add customizable columns to nearly any table.

Additional changes:

* respect the display_fqdn_for_hosts setting in the 'Name' column
* Breadcrumb link from host detail page is now a React link (when using the new hosts list page), so it loads faster

jeremylenz committed:
Fixes #37293 - Add user columns to hosts index
- add column registry
- add plugin documentation
- Make breadcrumb a React link
- Respect display_fqdn_for_hosts setting

ekohl commented on webpack/assets/javascripts/react_app/components/HostsIndex/Columns/core.js:
```diff
@@ -0,0 +1,63 @@
+  {
+    columnName: 'name',
+    title: __('Name'),
+    wrapper: ({ name }) => <Link to={`hosts/${name}`}>{name}</Link>,
```
I know today it's common to link to names, but I think it's fragile and we should link to IDs. Hostnames can change, IDs can't.

Also, and this is a much bigger issue, I think the REST API should provide `html_url` like GitHub does so the front end code doesn't have to guess. Just fill in verbatim what the API tells it. And that for every single resource.

parthaa replied:
Neat idea. I like it. However there are a quite a few pages hard coding the url's instead of relying on the API. Even if we decide that is bad and needs a change we'd have to address them en masse . May be we can file a redmine for that and track it as a separate issue

jeremylenz replied:
> I know today it's common to link to names, but I think it's fragile and we should link to IDs. Hostnames can change, IDs can't.

True, but it's also nice to see your hostname in the url bar when you're on the page. When we originally added this page it was linking to, say, `new/hosts/42` and that was a much worse experience IMO.

ekohl commented on webpack/assets/javascripts/react_app/components/HostsIndex/Columns/core.js:
```diff
@@ -0,0 +1,63 @@
+  {
+    columnName: 'hostgroup',
+    title: __('Host group'),
+    wrapper: hostDetails => hostDetails?.hostgroup_name,
```
AFAIK today the hostgroup is a link.

jeremylenz replied:
I can make it a link, no problem. For now I was just going off the mockup.

ekohl commented on webpack/assets/javascripts/react_app/components/HostsIndex/Columns/core.js:
```diff
@@ -0,0 +1,63 @@
+  {
+    columnName: 'owner',
+    title: __('Owner'),
+    wrapper: hostDetails => hostDetails?.owner_name,
```
I'm not sure if it's good or not, but owners can be users or groups. In some places we've used an icon to differentiate. Something to consider. I'm not sure today you **can** distinguish between the two in the REST API, but that could be a good reason to enhance the API.

ianballou replied:
Example:
![image](https://github.com/theforeman/foreman/assets/14796566/44e62c7c-5f1f-4a51-8cd0-b961a8644b99)

jeremylenz commented on webpack/assets/javascripts/react_app/components/ColumnSelector/ColumnSelector.js:
```diff
@@ -32,8 +33,10 @@ const ColumnSelector = props => {
+    setSaving(true);
     if (!hasPreference) {
-      await API.post(url, { name: 'hosts', columns: getColumnKeys() });
+      await API.post(url, { name: controller, columns: getColumnKeys() });
```
I'm not sure why "hosts" was hard-coded here, but this should make this component completely reusable.

jeremylenz replied:
This uses `API.post` directly, which is way outside the normal Foreman pattern of using Redux, but I wanted to get this functioning and just reuse the component without too many changes.

jeremylenz commented on webpack/assets/javascripts/react_app/components/ColumnSelector/ColumnSelector.js:
```diff
@@ -167,6 +171,8 @@ const ColumnSelector = props => {
               key="save"
               variant="primary"
+              isLoading={saving}
+              isDisabled={saving}
```
not sure why this isn't working

jeremylenz commented on webpack/assets/javascripts/react_app/components/ColumnSelector/helpers.js:
```diff
@@ -0,0 +1,87 @@
+  return false;
+};
+
+export const categoriesFromFrontendColumnData = ({
```
This builds the data object that is passed to the `ColumnSelector` component (and I think is passed down in more or less the same form to the Patternfly TreeView component.) I didn't change the data shape at all, so it should still work fine with the legacy page as well. It's a bit tricky because the legacy page mounts the component directly, so it can feed props to it directly from the backend, but we don't have that luxury here, so we need to "rebuild" this object.

jeremylenz commented on webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/Table/helpers.js:
```diff
@@ -21,13 +21,51 @@ export const getPageStats = ({ total, page, perPage }) => {
   return [columnNamesKeys, keysToColumnNames];
 };
+
+export const DEFAULT_USER_COLUMNS = [
+  'name',
+  'hostgroup',
+  'os_title',
+  'owner',
+  'last_report',
+];
```
These are the default columns as illustrated in the mockups.

jeremylenz commented:
Ready for initial review. Will squash and clean up the commits later.

ShimShtein commented:
A couple of comments here: 
1. I see this feature assumes the server will always return a superset of all the available columns that could possibly be displayed. 
2. Can we also extend the data retrieval mechanism to initiate additional requests when the main request comes back? This should solve the power status column and all columns that are not part of the original request
3. I think the current implementation actually deprecates pagelets and will require rewrite of all plugins that use them. Maybe we should at least for a while render the pagelets too?

jeremylenz commented:
> 1. I see this feature assumes the server will always return a superset of all the available columns that could possibly be displayed.

If the table preferences API returns a column that is not defined and registered, it will simply ignore that column. In this way you can have a column that is defined via a pagelet, but not in React, and it will work fine. This also allows incremental adoption among plugins.



> 2\. Can we also extend the data retrieval mechanism to initiate additional requests when the main request comes back? This should solve the power status column and all columns that are not part of the original request

Yes, I was thinking about doing something like that. I'd like to leave that for a follow-up if that's okay


> 3\. I think the current implementation actually deprecates pagelets and will require rewrite of all plugins that use them. Maybe we should at least for a while render the pagelets too?

This is just adding a way to extend React tables. For ERB pages you can still use pagelets, and in fact the pagelets still work fine on the legacy Hosts page. I suppose we could deprecate them, but as long as we still have ERB pages I'm thinking why not just have both!

ShimShtein commented:
> If the table preferences API returns a column that is not defined and registered, it will simply ignore that column. In this way you can have a column that is defined via a pagelet, but not in React, and it will work fine. This also allows incremental adoption among plugins.

If the page is built in React, how would pagelets be able to participate in table build, unless the react component would actually take into account the erb-generated HTML?.
My original comment was about the data source for column values. Like where do you take the equivalent of `@host` object from? I suppose it's an API call, something like `api/v2/hosts`. So to display a column you have to declare it in the table *and* make sure its data is returned in the API response itself.

>  I was thinking about doing something like that. I'd like to leave that for a follow-up if that's okay

Sounds good

> This is just adding a way to extend React tables. For ERB pages you can still use pagelets, and in fact the pagelets still work fine on the legacy Hosts page. I suppose we could deprecate them, but as long as we still have ERB pages I'm thinking why not just have both!

I was referring to hosts page specifically. Once we go with the react table, we won't support ERB generated content, hence we will deprecate pagelets for hosts. Of course in other pages the pagelets support will remain as is.

jeremylenz commented:
> I was referring to hosts page specifically. Once we go with the react table, we won't support ERB generated content, hence we will deprecate pagelets for hosts. Of course in other pages the pagelets support will remain as is.

sure, makes sense to me! We'll have to decide down the road what the right timeline is for deprecating pagelets. Right now the plan for new All Hosts page is

1. Build Top 7 actions on the new hosts page (Delete, HG, Build, LCE, Content source, Errata, Packages)
2. Continue building more actions (maybe?) starting with the most requested ones
3. Once we feel like we have enough functionality on the new page, switch the default to the new page
4. (At some point after that) deprecate the old page
5. (At some point after that) remove the old page

So we've still got quite a bit of time before all that.

ianballou commented:
I noticed that, for some reason, the Owner column doesn't show the full name if you hover over it:

[Screencast from 2024-04-10 14-21-11.webm](https://github.com/theforeman/foreman/assets/14796566/8b5fb5b4-66fa-457f-9d13-85208d676824)

Edit: The video appears to be horribly corrupt, but just replace the weird graphical artifacts with the hover boxes in your head :)

ianballou commented:
Unlike on the legacy page, I'm able completely empty out the columns:
![image](https://github.com/theforeman/foreman/assets/14796566/86ad7af9-ac43-46bf-a69e-171650e79d59)

When this happens, I see an error in the logs but the page just hangs. On the legacy page, the name must be there.

ianballou reviewed (commented):
Overall working well, left comments on some small things I found. Most comments are above, decided to make an official review so I'm re-requestable.

parthaa commented:
Looking great!. Couple of things I noticed
1) Should disable unchecking the name column to be consistent with old ui.
2) On the initial load of the page I see a bunch of columns show up before I see the word "Loading" and then the columns get changed. Is there any thing you can do to improve that behavior. Not a big deal.

jeremylenz commented:
> I noticed that, for some reason, the Owner column doesn't show the full name if you hover over it:

I reproduced this. Crazy that the issue is only with the Owner column. I can't for the life of me figure out why it's happening. https://v4-archive.patternfly.org/v4/components/table says I should be able to override this with a "tooltip" prop on the `<Th>`, but that seems to work only if the column isn't sortable. Sortable columns just seem to not have a controllable tooltip. It's probably a Patternfly bug, which won't get fixed since PF4 is EOL now.

jeremylenz commented:
Added icons for Owner column :)

![image](https://github.com/theforeman/foreman/assets/22042343/a6568ba7-0959-4666-a97d-c8f6e4d54393)

jeremylenz commented:
1. Name column is now required (disabled & checked)
2. Added icons for Owner column
3. Suppress the display of any columns while the table is loading (should fix your issue @parthaa )
4. squashed commits

jeremylenz commented:
updated snapshots.

parthaa commented:
ACK all my concerns have been addressed. PR good to merge

parthaa reviewed (approved):
ACK

ianballou reviewed (approved):
Gave it some more testing, confirmed all of the prior concerns were dealt with besides the Owner issue that may be a PF bug. Looking good to me!

Also compared with the current page, the only column functionalities that differ are the ones that need futher work (Power and Virtual).
